Overwrite Rule Properties Without Excluding the Rule

tis-dm-webfactory opened the pull request:
Add possibility to modify rule properties via config file without defining a completely new rule like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<ruleset name="pcsg-generated-ruleset"
         xmlns="http://pmd.sf.net/ruleset/1.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://pmd.sf.net/ruleset/1.0.0 http://pmd.sf.net/ruleset_xml_schema.xsd"
         xsi:noNamespaceSchemaLocation="http://pmd.sf.net/ruleset_xml_schema.xsd">
    <description>Created with the PHP Coding Standard Generator.
        http://edorian.github.com/php-coding-standard-generator/
    </description>

    <rule ref="rulesets/design.xml" />
    <rule name="CouplingBetweenObjects">
        <properties>
            <property name="maximum" value="20" />
        </properties>
    </rule>
</ruleset>
```

Fix #39

Commit:
add possibility to modify rule properties via config file without defining a completely new rule

kylekatarnls committed:
Fix PHP 5.3 compatibility

kylekatarnls commented:
Hi thanks, I like this feature a lot. I fixed the PHP 5.3 compatibility in your PR (please pull the change). Could you now add a unit test for that?

And it would be perfect to add a note about this in https://github.com/phpmd/phpmd/blob/master/src/site/rst/documentation/creating-a-ruleset.rst

So other users will know this exists.

Commit:
add unit tests for new config feature

Commit:
add some doc

tis-dm-webfactory committed:
Update creating-a-ruleset.rst

Commit:
add some documentation
add tests

Commit:
Merge branch 'master' of git://github.com/tis-dm-webfactory/phpmd

tis-dm-webfactory commented:
hi Kyle,

yeah sure - I've added two (simple) tests and also wrote a paragraph about how this feature works.
I've pushed it in our repo - if you need another pull request just let me know.

kylekatarnls committed:
Fix underline alignment

kylekatarnls reviewed (dismissed):
Very nice!

MarkVaughn reviewed (dismissed):
Noice

kylekatarnls commented:
@ravage84 A last word on this before merging?

ravage84 commented on src/main/php/PHPMD/RuleSetFactory.php:
```diff
@@ -436,6 +442,37 @@ private function parseRuleReferenceNode(RuleSet $ruleSet, \SimpleXMLElement $rul
     }
 
+    /**
+     * modifyExistingRuleset
```
Please use an actual phrase, you can use the explanation two lines below.

tis-dm-webfactory replied:
I've change the comment, please see the latest update

ravage84 commented on src/main/php/PHPMD/RuleSetFactory.php:
```diff
@@ -436,6 +442,37 @@ private function parseRuleReferenceNode(RuleSet $ruleSet, \SimpleXMLElement $rul
+
+        $ruleName = (string) $attributes['name'];
+
+        // find existing rule
```
A method like this often indicates that it is better to have a sub method with a self explanatory name (plus doc block).

tis-dm-webfactory replied:
I've extracted searching for a rule to a method - please see latest commit

ravage84 commented on src/site/rst/documentation/creating-a-ruleset.rst:
```diff
@@ -246,6 +246,49 @@ __ /rules/cleancode.html
       </rule>
   </ruleset>
 
+Changing properties in a rule set for multiple rules
```
Important words in titles can be written in capital letters.

tis-dm-webfactory replied:
I'm sorry, but I can't see anything wrong here. The titles look similar throughout the whole file ;-)

kylekatarnls replied:
Good for me.

ravage84 commented on src/test/php/PHPMD/RuleSetFactoryTest.php:
```diff
@@ -707,6 +707,71 @@ public function testIfGettingRuleFilePathExcludeUnreadablePaths()
     }
 
+    /**
+     * testCreateRuleSetsWithoutRuleReferenceThatOverwritesSettings
```
Please use full sentences. Otherwise this has no added benefit.

tis-dm-webfactory replied:
I've change the comment, please see the latest update

ravage84 commented on src/test/php/PHPMD/RuleSetFactoryTest.php:
```diff
@@ -707,6 +707,71 @@ public function testIfGettingRuleFilePathExcludeUnreadablePaths()
+    }
+
+    /**
+     * testCreateRuleSetsWithoutRuleReferenceThatOverwritesSettingsButDoesntAddProperties
```
Please use full sentences. Otherwise this has no added benefit.

tis-dm-webfactory replied:
I've change the comment, please see the latest update

ravage84 commented on src/test/resources/files/rulesets/refset5.xml:
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<ruleset name="Third Test RuleSet"
```
Fifth

ravage84 commented on src/test/resources/files/rulesets/refset5.xml:
```diff
@@ -0,0 +1,27 @@
+         xsi:schemaLocation="http://pmd.sf.net/ruleset/1.0.0 http://pmd.sf.net/ruleset_xml_schema.xsd"
+         xsi:noNamespaceSchemaLocation="http://pmd.sf.net/ruleset_xml_schema.xsd">
+
+    <description>First description...</description>
```
May be we can use the description to describe what is going on in this rule set file

ravage84 commented on src/test/resources/files/rulesets/set3.xml:
```diff
@@ -0,0 +1,46 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<ruleset name="Second Test RuleSet"
```
Third

ravage84 commented on src/test/resources/files/rulesets/set3.xml:
```diff
@@ -0,0 +1,46 @@
+		 xsi:schemaLocation="http://pmd.sf.net/ruleset/1.0.0 http://pmd.sf.net/ruleset_xml_schema.xsd"
+		 xsi:noNamespaceSchemaLocation="http://pmd.sf.net/ruleset_xml_schema.xsd">
+
+	<description>Second description...</description>
```
May be we can use the description to describe what is going on in this rule set file

ravage84 reviewed (changes requested):
If I understand this PR correctly, it adds the possibility to overwrite the properties of rules without excluding them, once they are added through a rule set, correct?

Commit:
modified/extended documentation for 'change properties feature'

Commit:
extended documentation for tests of 'change properties feature'
fixed errors from copy/paste in test rulesets
added description to test ruleset to say what's going on

Commit:
refactoring to (hopefully) match request changes:
extract searching for a rule to a method
improve/modify comments for added methods

Commit:
remove return type - try to resolve problem in phar build

tis-dm-webfactory commented:
> If I understand this PR correctly, it adds the possibility to overwrite the properties of rules without excluding them, once they are added through a rule set, correct?

absolutely correct ;-)

tis-dm-webfactory commented:
@ravage84
@MarkVaughn 
@kylekatarnls 
do you need any more changes? there's been no more comments since my last update.

kylekatarnls reviewed (dismissed):
It's alright for me.

Commit:
Merge branch 'master' of git://github.com/phpmd/phpmd

tvbeek reviewed (dismissed):
Nice PR, this made it more easy to change the rules for your need.

Thanks

ravage84 commented on src/site/rst/documentation/creating-a-ruleset.rst:
```diff
@@ -246,6 +246,51 @@ __ /rules/cleancode.html
+          My custom rule set that checks my code...
+      </description>
+
+      <!--
```
The previous examples build upon the one example before them. Should we keep that narrative style?

kylekatarnls replied:
I don't like neither. I will add some `if` and `for instance` to make it clear it's just an example.

kylekatarnls replied:
In fact, it's used very widely, we should handle this in a next PR. And keep it consistent like it is for now.

ravage84 commented on src/main/php/PHPMD/RuleSetFactory.php:
```diff
@@ -344,6 +344,12 @@ private function parseSingleRuleNode(RuleSet $ruleSet, \SimpleXMLElement $ruleNo
             $fileName = str_replace(array('\\', '_'), '/', $className) . '.php';
         }
 
+        if ($className === '') {
```
The intend of this coude section could be communicated better. May be a comment is necessary?

ravage84 committed:
CS fixes

ravage84 committed:
CS fixes

ravage84 committed:
Improve variable name

ravage84 reviewed (dismissed):
@tis-dm-webfactory thank you for your work. It won't take much to make this mergable.
Can you also please rebase it upon our current master?

ravage84 commented:
I think this PR resolves https://github.com/phpmd/phpmd/issues/39

kylekatarnls reviewed (dismissed):
What about merging and cleanup old documentation in an other PR?

kylekatarnls reviewed (dismissed):
I moved the early-return earlier in the method and pulled master.

Side note: It's a lot of commits now. I propose to squash this PR.

kylekatarnls committed:
Merge branch 'master' of https://github.com/phpmd/phpmd into master-override-rule

kylekatarnls committed:
Exit parseSingleRuleNode as early as possible

ravage84 commented:
@kylekatarnls 

> What about merging and cleanup old documentation in an other PR?

Can we address https://github.com/phpmd/phpmd/pull/664#discussion_r420938937 before doing so?

> Side note: It's a lot of commits now. I propose to squash this PR.

I agree.

kylekatarnls commented on src/test/php/PHPMD/RuleSetFactoryTest.php:
```diff
@@ -707,6 +707,71 @@ public function testIfGettingRuleFilePathExcludeUnreadablePaths()
+    }
+
+    /**
+     * Check if properties are changed, but no additional properties are added to rules
```
This test is not useless, it fails on the current version of RuleSetFactory and it succeed with this change. I will update it to make it a bit more short and clear.

kylekatarnls commented:
Addressed https://github.com/phpmd/phpmd/pull/664/commits/8d296f83ed50e39c1fa796628a00c07f76f45061

If I well understand, this does not change the intent of the test.

kylekatarnls committed:
Simplify the unit test for OutOfBoundsException

ravage84 commented:
@kylekatarnls 

> Addressed [8d296f8](https://github.com/phpmd/phpmd/commit/8d296f83ed50e39c1fa796628a00c07f76f45061)
> 
> If I well understand, this does not change the intent of the test.

I figured it out:

The point of this test is that the property ``test`` in rule ``RuleThreeInFifthRuleSet`` from [src/test/resources/files/rulesets/refset5.xml](https://github.com/phpmd/phpmd/pull/664/files#diff-6ff82293398c390e7df1f1d334240c22R25) should not add this property to that rule as it didn't exist before in the original rule set [src/test/resources/files/rulesets/set3.xml](https://github.com/phpmd/phpmd/pull/664/files#diff-7405741fb106eec9cc487ac854275b28R36-R45).

The current test doesn't realy cover this as it would fail at the [first retrieval of the property](https://github.com/phpmd/phpmd/pull/664/files#diff-f7bef854eeb9f7fb07bbd08f2c75ec82R767) because [the names of the properties for the first two rules](https://github.com/phpmd/phpmd/pull/664/files#diff-6ff82293398c390e7df1f1d334240c22R15-R20) in the rule set are ``test1`` and ``test2`` and not ``test``.

To cover this case we either have to look at the full list of properties (which would also be a more stable test) or just fetch the third rule.

kylekatarnls commented on src/site/rst/documentation/creating-a-ruleset.rst:
```diff
@@ -246,6 +246,51 @@ __ /rules/cleancode.html
+Changing properties in a rule set for multiple rules
+====================================================
+
+We want to use the basic codesize ruleset, but change parameter values for
```
I'll remove the previous chapter.

kylekatarnls committed:
Remove obsolete chapter about changing individual property

kylekatarnls commented on src/main/php/PHPMD/RuleSetFactory.php:
```diff
@@ -436,6 +446,59 @@ private function parseRuleReferenceNode(RuleSet $ruleSet, \SimpleXMLElement $rul
+     * @return AbstractRule
+     * @throws \RuntimeException if no matching rule was found
+     */
+    private function findRuleInRulesetByName(RuleSet $ruleSet, string $ruleName)
```
string type hint is not possible in PHP 5

kylekatarnls commented on src/main/php/PHPMD/RuleSetFactory.php:
```diff
@@ -436,6 +446,59 @@ private function parseRuleReferenceNode(RuleSet $ruleSet, \SimpleXMLElement $rul
+        $rules = $ruleSet->getRules();
+
+        // Find existing rule
+        while ($rules->valid()) {
```
This is an iterator, it can be iterated simply using `foreach ($rulesSet as $rule)` with no calls to `->getRules()` `->valid()` `->current()` or `->next()`

kylekatarnls reviewed (changes requested):
The feature is actually not compatible with PHP 5. The unit tests only pass because the test trigger 0 assertions in PHP 5.

kylekatarnls committed:
Use a typed exception for RuleNotFound

So the catch() does not catch unwanted exception

kylekatarnls commented:
Actually the test does not perform any assertions neither in PHP 7 because the `array_key_exists($name, $expected)` is always `false`.

kylekatarnls committed:
Fix trailing commas

kylekatarnls committed:
Fix iterators iteration

kylekatarnls committed:
Remove PHP-7 only type hint

kylekatarnls committed:
Test the whole set of properties

kylekatarnls commented:
I fixed the test to make it check what it intended to and sadly it fails (0 instead of 42 in test1 property).

kylekatarnls commented on src/site/rst/documentation/creating-a-ruleset.rst:
```diff
@@ -209,39 +204,24 @@ __ /rules/cleancode.html
-      <rule ref="rulesets/codesize.xml/CyclomaticComplexity">
-          <priority>1</priority>
+      <rule ref="rulesets/codesize.xml"/>
+      <rule name="ExcessiveParameterList">
```
I would actually prefer that the rule would be inside the ` <rule ref="rulesets/codesize.xml">...</rule>`

kylekatarnls replied:
After a second thinking, it's fine, we should not have the same rule name in 2 different rule-sets and it's easier for the user if he/she doesn't need to know the rule-set of every rule.

kylekatarnls committed:
Fix XML spacing in example

kylekatarnls commented:
I checked, It's just a problem with the test. The feature works properly with the example given in the new documentation.

kylekatarnls committed:
Fix unit test using correct rule names

kylekatarnls commented:
Test fixed. It was a naming problem: RuleOneInThirdRuleSet instead of RuleOneInFifthRuleSet as it refers to the set3.xml.

kylekatarnls committed:
Fix out-of-bounds test

kylekatarnls reviewed (dismissed):
Tests fixed.
https://travis-ci.org/github/phpmd/phpmd/builds/690347990

tvbeek commented:
@tis-dm-webfactory I'm realizing that this is an old PR. Are you able to rebase it to handle the conflicts and let it run the tests with the GitHub actions?

AJenbo committed:
Merge branch '3.x' into rule-overwrite

AJenbo committed:
Merge branch '3.x' of github.com:phpmd/phpmd into tis

AJenbo committed:
Fix line length violation and clean up test reflection usage

AJenbo committed:
Restore exclude+re-include docs alongside new simpler overwrite syntax

AJenbo committed:
Support property overwriting for JSON/YAML/PHP config formats, fix JSON
parsing bug
